### PR TITLE
Internal refactors as a path toward #18062

### PR DESCRIPTION
**Pull request by freben**
#18062 does too many things. The actual effective changes tend to get lost among the refactors that were necessary to support the new feature. So here's a PR with the (maybe?) non-controversial internal refactors, which arguably have 🧹 value on their own.

Overview of this PR:

- No externally visible changes, no functional changes neither in terms of stitching nor anything else
- The `Stitcher` class is broken apart into a `Stitcher` interface and a `DefaultStitcher` implementation class
- Some more features are moved to the `database` folder as a continuation of previous efforts:
  - `deleteOrphanedEntities` (well it was already there, but now also stripped from `DefaultProcessingDatabase`)
  - `markForStitching` (used to be hidden in `DefaultProcessingDatabase`)
  - `performStitching` (the meat of what was in `Stitcher`)
  - `buildEntitySearch`
- Documented `DbRefreshStateRow`
- `DefaultCatalogProcessingEngine` no longer implements `CatalogProcessingEngine` since there is no reason to; the latter only has start and stop, and now also will start affecting the stitcher. The similar naming is nowadays mostly a historical artifact.
- Broke out the `progressTracker` for the stitcher into a file of its own so we more easily can deal with the prom / otel duality and iterate on the metrics

### CODE REVIEW

File: plugins/catalog-backend/src/database/operations/util/deleteOrphanedEntities.ts [hunk 1]
```diff
@@ -72,26 +75,17 @@ export async function deleteOrphanedEntities(options: {
...
-      .whereIn('entity_id', orphanRelationIds);
-    await tx
-      .table<DbRefreshStateRow>('refresh_state')
-      .update({
-        result_hash: 'orphan-relation-deleted',
-        next_update_at: tx.fn.now(),
-      })
-      .whereIn('entity_id', orphanRelationIds);
+    // Mark all of the things that the orphans had relations to for stitching
+    await markForStitching({
```
Comment (alde) [inline plugins/catalog-backend/src/database/operations/util/deleteOrphanedEntities.ts:84]: maybe I'm misunderstanding, but wouldn't this introduce the deferred stitching?

Comment (freben) [reply plugins/catalog-backend/src/database/operations/util/deleteOrphanedEntities.ts:84]: No, even in the old system there was a way to mark something for stitching _kind of_, which is what this retains. No functional changes as such. What it does is to mark the entity for both re-processing and re-stitching.

In the later PR, it'll have an if statement in that function doing it the old way or the more efficient new way, depending on settings

**Review by alde — APPROVED** [review]

File: .changeset/polite-donuts-fail.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+---
+'@backstage/plugin-catalog-backend': patch
+---
+
+Internal refactors as a path toward #18062
```
Comment (Rugvip) [inline .changeset/polite-donuts-fail.md:5]: Probably better to explain here what #18062 is with a few words :p

**Review by Rugvip — APPROVED** [review]
🎉

File: .changeset/polite-donuts-fail.md (hunk 2 above)
Comment (freben) [reply .changeset/polite-donuts-fail.md]: aight, updated

**Pull request merged**